### PR TITLE
feat: GET /users/me

### LINKED ISSUE #77: `GET /users/me`
```
Response 200
{
  "id": <id>,
  "nickname": string,
  "email": string
}
```

### PR DESCRIPTION
**Pull request by lohas1107**
## Why need this change? / Root cause: 
- As a user, I want to see my profile
## Changes made:
- User related
## Test Scope / Change impact:
- UserController
## Issue
- resolve #77

### CODE REVIEW

File: spring/src/main/kotlin/tw/waterballsa/gaas/spring/controllers/UserController.kt [hunk 1]
```diff
@@ -1,15 +1,23 @@
...
+import org.springframework.security.oauth2.core.oidc.user.OidcUser
 import org.springframework.web.bind.annotation.GetMapping
 import org.springframework.web.bind.annotation.PathVariable
+import org.springframework.web.bind.annotation.RequestMapping
 import org.springframework.web.bind.annotation.RestController
 import tw.waterballsa.gaas.application.usecases.GetUserUseCase
 import tw.waterballsa.gaas.domain.User
 
 @RestController
+@RequestMapping("/users")
```
Comment (Wally5077) [inline spring/src/main/kotlin/tw/waterballsa/gaas/spring/controllers/UserController.kt:13]: Excellent!!!

File: spring/src/test/kotlin/tw/waterballsa/gaas/spring/it/AbstractSpringBootTest.kt [hunk 2]
```diff
@@ -24,4 +29,22 @@ abstract class AbstractSpringBootTest {
...
+            "token",
+            Instant.now(),
+            Instant.now().plusSeconds(60),
+            mapOf(
+                "sub" to mockUser.id!!.value,
+                "name" to mockUser.nickname,
+                "email" to mockUser.email
+            )
+        )
+        return DefaultOidcUser(emptyList(), idToken)
```
Comment (Wally5077) [inline spring/src/test/kotlin/tw/waterballsa/gaas/spring/it/AbstractSpringBootTest.kt:48]: Cool

File: spring/src/test/kotlin/tw/waterballsa/gaas/spring/it/controllers/UserControllerTest.kt [hunk 3]
```diff
@@ -12,40 +12,70 @@ import tw.waterballsa.gaas.application.repositories.UserRepository
 import tw.waterballsa.gaas.domain.User
 import tw.waterballsa.gaas.spring.it.AbstractSpringBootTest
 
-
-@AutoConfigureMockMvc(addFilters = false)
 class UserControllerTest @Autowired constructor(
     val userRepository: UserRepository,
 ) : AbstractSpringBootTest() {
 
+    var testUser = User(User.Id("1"), "user@example.com", "winner5566")
```
Comment (Wally5077) [inline spring/src/test/kotlin/tw/waterballsa/gaas/spring/it/controllers/UserControllerTest.kt:19]: 變數放到父類

**Review by Wally5077 — CHANGES_REQUESTED** [review]
Review done.

**Review by frankvicky — APPROVED** [review]
LGTM

File: spring/src/test/kotlin/tw/waterballsa/gaas/spring/it/controllers/UserControllerTest.kt (hunk 3 above)
Comment (lohas1107) [reply spring/src/test/kotlin/tw/waterballsa/gaas/spring/it/controllers/UserControllerTest.kt]: @Wally5077 已放到父類

File: spring/src/main/kotlin/tw/waterballsa/gaas/spring/controllers/UserController.kt [hunk 4]
```diff
@@ -1,15 +1,23 @@
...
 class UserController(
     private val getUserUseCase: GetUserUseCase
 ) {
-    @GetMapping("/users/{id}")
+    @GetMapping("/{id}")
     fun getUser(@PathVariable id: String): User = getUserUseCase.execute(User.Id(id))
+
+    @GetMapping("/me")
+    fun getUserMe(@AuthenticationPrincipal principal: OidcUser): User =
+        getUserUseCase.execute(User.Id(principal.subject))
```
Comment (ricksu978) [inline spring/src/main/kotlin/tw/waterballsa/gaas/spring/controllers/UserController.kt:22]: 這邊要改成 Jwt

**Review by ricksu978 — CHANGES_REQUESTED** [review]

**Review by ricksu978 — APPROVED** [review]
LGTM

File: application/src/main/kotlin/tw/waterballsa/gaas/application/usecases/GetUserMeUseCase.kt [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+@Named
+class GetUserMeUseCase(
+    private val userRepository: UserRepository,
+) {
+    fun execute(request: Request, presenter: Presenter) {
+        with(request) {
+            when (val user = userRepository.findByEmail(email)) {
+                null -> throw notFound(User::class).identifyBy("email", email)
+                else -> presenter.present(user)
+            }
```
Comment (Wally5077) [inline application/src/main/kotlin/tw/waterballsa/gaas/application/usecases/GetUserMeUseCase.kt:17]:
```kotlin
val user = userRepository.findByEmail(email)?: notFound(User::class).identifyBy("email", email)
presenter.present(user)
```

File: spring/src/test/kotlin/tw/waterballsa/gaas/spring/it/AbstractSpringBootTest.kt [hunk 6]
```diff
@@ -22,7 +22,7 @@ abstract class AbstractSpringBootTest {
     @Autowired
     private lateinit var objectMapper: ObjectMapper
 
-    protected final val mockUser: User = User(
+    protected var mockUser: User = User(
```
Comment (Wally5077) [inline spring/src/test/kotlin/tw/waterballsa/gaas/spring/it/AbstractSpringBootTest.kt:25]: Q：為什麼需要 var？

File: spring/src/test/kotlin/tw/waterballsa/gaas/spring/it/controllers/UserControllerTest.kt [hunk 7]
```diff
@@ -24,28 +21,35 @@ class UserControllerTest @Autowired constructor(
     }
 
     @Test
-    fun givenUserCreated_whenGetUser_thenGetUserSuccessfully() {
-        val user = User(User.Id("1"), "test@mail.com", "winner5566")
-        givenUserCreated(user)
-        findUserById("1").thenGetUserSuccessfully(user)
+    fun givenUserCreated_whenGetUserMe_thenShouldReturnUserInfo() {
```
Comment (Wally5077) [inline spring/src/test/kotlin/tw/waterballsa/gaas/spring/it/controllers/UserControllerTest.kt:24]: **givenUserHasLoggedIn_whenGetUserSelf_thenGetUserSuccessfully**

File: spring/src/test/kotlin/tw/waterballsa/gaas/spring/it/controllers/UserControllerTest.kt [hunk 8]
```diff
@@ -24,28 +21,35 @@ class UserControllerTest @Autowired constructor(
...
+        givenUserCreated()
+            .whenGetUserMe()
```
Comment (Wally5077) [inline spring/src/test/kotlin/tw/waterballsa/gaas/spring/it/controllers/UserControllerTest.kt:26]: 對應測試案例名稱

File: spring/src/main/kotlin/tw/waterballsa/gaas/spring/controllers/UserController.kt [hunk 9]
```diff
@@ -1,15 +1,28 @@
...
 @RestController
+@RequestMapping("/users")
 class UserController(
-    private val getUserUseCase: GetUserUseCase
+    private val getUserMeUseCase: GetUserMeUseCase
 ) {
-    @GetMapping("/users/{id}")
-    fun getUser(@PathVariable id: String): User = getUserUseCase.execute(User.Id(id))
+    @GetMapping("/me")
+    fun getUserMe(@AuthenticationPrincipal principal: Jwt): GetUserMeViewModel {
```
Comment (Wally5077) [inline spring/src/main/kotlin/tw/waterballsa/gaas/spring/controllers/UserController.kt:19]: **getUser**

File: spring/src/main/kotlin/tw/waterballsa/gaas/spring/controllers/UserController.kt [hunk 10]
```diff
@@ -1,15 +1,28 @@
...
+        val request = principal.toRequest()
+        val presenter = GetUserMePresenter()
```
Comment (Wally5077) [inline spring/src/main/kotlin/tw/waterballsa/gaas/spring/controllers/UserController.kt:21]: **GetUserPresenter**

File: spring/src/main/kotlin/tw/waterballsa/gaas/spring/controllers/UserController.kt [hunk 11]
```diff
@@ -1,15 +1,28 @@
...
+import tw.waterballsa.gaas.application.usecases.GetUserMeUseCase
 import tw.waterballsa.gaas.domain.User
+import tw.waterballsa.gaas.spring.controllers.presenter.GetUserMePresenter
+import tw.waterballsa.gaas.spring.controllers.viewmodel.GetUserMeViewModel
 
 @RestController
+@RequestMapping("/users")
 class UserController(
-    private val getUserUseCase: GetUserUseCase
+    private val getUserMeUseCase: GetUserMeUseCase
```
Comment (Wally5077) [inline spring/src/main/kotlin/tw/waterballsa/gaas/spring/controllers/UserController.kt:16]: **GetUserUseCase**

File: spring/src/test/kotlin/tw/waterballsa/gaas/spring/it/controllers/UserControllerTest.kt [hunk 12]
```diff
@@ -24,28 +21,35 @@ class UserControllerTest @Autowired constructor(
...
+            .thenGetUserSuccessfully()
     }
 
     @Test
-    fun givenUserNotCreated_whenGetUser_thenUserNotFound() {
-        findUserById("0").thenUserNotFound()
+    fun givenUserNotCreated_whenGetUserMe_thenUserNotFound() {
```
Comment (Wally5077) [inline spring/src/test/kotlin/tw/waterballsa/gaas/spring/it/controllers/UserControllerTest.kt:31]: **givenUserDoesNotLogIn_whenGetUserSelf_thenUserNotFound**

**Review by Wally5077 — CHANGES_REQUESTED** [review]
**Review done.**

**Review by Wally5077 — APPROVED** [review]
**LGTM**

**Pull request merged**